File: algorithms/blind75/2/climbing_stairs.py
# binary exponantiation of the Fibonacci matrix
class Solution:
    def climbStairs(self, n: int) -> int:
        F = [[1, 1], [1, 0]] # F^n = [ [F_(n+1), F_n], [F_n, F_(n-1)] ]
        return self.mat_pow(F, n)[0][0]
    
    def mat_mult(self, A, B):
        C = [[0, 0], [0, 0]]
        for i in range(len(A)):
            for j in range(len(B[0])):
                for k in range(len(A[0])):
                    C[i][j] += A[i][k] * B[k][j]
        print(C)
        return C
    
    def mat_pow(self, A, n):
        B = [[1, 0], [0, 1]]
        k = n.bit_length() - 1
        while k >= 0:
            B = self.mat_mult(B, B)
            if n & (1 << k):
                B = self.mat_mult(B, A)
            k -= 1
        return B

"""
A^n, n = 2^k + ... + 2^m = 10001, 101011, ..
"""

File: algorithms/blind75/2/climbing_stairs.cpp
// it was Fibonacci and missed it
class Solution {
public:
    int climbStairs(int n) {
        int ans = 1; // all 1's case

        for(int i = 1; i <= n/2; ++i) {
            unsigned long long k = 1;
            int l = max(n-2*i, i);
            for(int j = l+1, mi = 1, mv = 1; j <= n-i; ++j) { // calculate (n - i)!/(i!(n-2*i)!)
                if (mv == 1 && mi < n-i-l) mv = ++mi;
                k *= j;
                if (mv != 1) {
                    int g = gcd(k, mv);
                    mv /= g;
                    k /= g;
                }
            }
            // for(int j = l+1; j <= n-i; ++j) k *= j;
            // for(int j = 1; j <= n-i-l; ++j) k /= j;
            ans += k;
        }
        
        return ans;
    }
};

/*
    n = 2*i + m, i = 0..(n/2)
    for each i, a climbStair is made by successive choices from i 2's and m 1's which count in total as (i+m)!/(i!m!)
*/
